unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,699 @@
+
+        Some((cid, vec))
+    }
+
+    fn flush_buffered_links(&mut self, finishing: bool) -> Vec<(Cid, Vec<u8>)> {
+        self.collector
+            .flush_links(&mut self.unflushed_links, finishing)
+    }
+
+    /// Test helper for collecting all of the produced blocks; probably not a good idea outside
+    /// smaller test cases. When `amt` is zero, the whole content is processed at the speed of
+    /// chunker, otherwise `all_content` is pushed at `amt` sized slices with the idea of catching
+    /// bugs in chunkers.
+    #[cfg(test)]
+    fn collect_blocks(mut self, all_content: &[u8], mut amt: usize) -> Vec<(Cid, Vec<u8>)> {
+        let mut written = 0;
+        let mut blocks_received = Vec::new();
+
+        if amt == 0 {
+            amt = all_content.len();
+        }
+
+        while written < all_content.len() {
+            let end = written + (all_content.len() - written).min(amt);
+            let slice = &all_content[written..end];
+
+            let (blocks, pushed) = self.push(slice);
+            blocks_received.extend(blocks);
+            written += pushed;
+        }
+
+        let last_blocks = self.finish();
+        blocks_received.extend(last_blocks);
+
+        blocks_received
+    }
+}
+
+fn render_and_hash(flat: &FlatUnixFs<'_>) -> (Cid, Vec<u8>) {
+    // TODO: as shown in later dagger we don't really need to render the FlatUnixFs fully; we could
+    // either just render a fixed header and continue with the body OR links, though the links are
+    // a bit more complicated.
+    let mut out = Vec::with_capacity(flat.get_size());
+    let mut writer = Writer::new(&mut out);
+    flat.write_message(&mut writer)
+        .expect("unsure how this could fail");
+    let mh = multihash::wrap(multihash::Code::Sha2_256, &Sha256::digest(&out));
+    let cid = Cid::new_v0(mh).expect("sha2_256 is the correct multihash for cidv0");
+    (cid, out)
+}
+
+/// Chunker strategy
+#[derive(Debug, Clone)]
+pub enum Chunker {
+    /// Size based chunking
+    Size(usize),
+}
+
+impl std::default::Default for Chunker {
+    /// Returns a default chunker which matches go-ipfs 0.6
+    fn default() -> Self {
+        Chunker::Size(256 * 1024)
+    }
+}
+
+impl Chunker {
+    fn accept<'a>(&mut self, input: &'a [u8], buffered: &[u8]) -> (&'a [u8], bool) {
+        use Chunker::*;
+
+        match self {
+            Size(max) => {
+                let l = input.len().min(*max - buffered.len());
+                let accepted = &input[..l];
+                let ready = buffered.len() + l >= *max;
+                (accepted, ready)
+            }
+        }
+    }
+
+    fn size_hint(&self) -> usize {
+        use Chunker::*;
+
+        match self {
+            Size(max) => *max,
+        }
+    }
+}
+
+/// Collector or layout strategy
+#[derive(Debug, Clone)]
+pub enum Collector {
+    /// Balanced trees.
+    Balanced(BalancedCollector),
+}
+
+impl std::default::Default for Collector {
```

Comment:
doesn't this just work as a `derive` on `Collector`?

Reply:
It's an enum; it doesn't matter that it has only a single variant.

Reply:
ah yeah; it probably makes sense future-proof wise, when more variants could be added